Feature-proto-scara

joachimhgg opened the pull request:
**Description**

Add SCARA Robot PROTO to Webots. For the moment, the controller example is using the supervisor to mimic a suction cup, which should be replaced in another PR.

ad-daniel commented on docs/reference/changelog-r2022.md:
```diff
@@ -10,6 +10,7 @@ Released on XX, XXth, 2022.
     - Added two new Robot API functions, `wb_robot_step_begin` and `wb_robot_step_end`, to optimize computer intensive controllers ([#4107](https://github.com/cyberbotics/webots/pull/4107)).
   - New Objects:
     - Added some objects on the hospital theme: hospital bed, drip stand, medicine bottle, hand sanitizer, curtain, photo frame, flower pot, emergency exit sign and a fabric appearance ([#4166](https://github.com/cyberbotics/webots/pull/4166)).
+  - Added a [SCARA Epson T6](../guide/scara_epson_t6.md) PROTO and a demo showing it moving fruits in a food industry ([#4192](https://github.com/cyberbotics/webots/pull/4192).
```
Seems more clear to me, what do you think?

ad-daniel commented on projects/robots/epson/ScaraT6/worlds/industrial_example.wbt:
```diff
@@ -0,0 +1,514 @@
+#VRML_SIM R2022b utf8
+WorldInfo {
+  info [
+    "Universal Robot UR3e, UR5e and UR10e grasping cans using Robotiq 3F grippers."
```
Needs to be updated

ad-daniel commented on projects/robots/epson/ScaraT6/worlds/industrial_example.wbt:
```diff
@@ -0,0 +1,514 @@
+}
+Apple {
+  translation -0.0126866 -0.903111 0.45
+  rotation 0.9999999999959461 1.733409999992973e-06 -2.2590199999908423e-06 -1.3089953071795861
```
```suggestion
  rotation 1 0 0 -1.3089953071795861
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0477 -0.0494 0.4091
+          rotation 0 -1 0 -1.0472053071795866
```
```suggestion
          rotation 0 1 0 1.0472053071795866
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0921 -0.0493 0.3583
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0934 -0.05565 0.2079
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0934 -0.05565 0.2382
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0932 0.05565 0.2383
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0934 0.05565 0.2078
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0934 0.05565 -0.0014
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0934 -0.05569 -0.0014
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0921 0.0498 0.3583
+          rotation 0 1 0 0
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+        }
+        CapScrew {
+          translation -0.0477 0.0496 0.4091
+          rotation 0 -1 0 -1.0472053071795866
```
```suggestion
          rotation 0 1 0 1.0472053071795866
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+                        }
+                        CapScrew {
+                          translation -0.3091 -0.01371 0.0509
+                          rotation 0 -1 0 -5.307179586466759e-06
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+                        }
+                        CapScrew {
+                          translation -0.3091 0.0131 0.0509
+                          rotation 0 -1 0 -5.307179586466759e-06
```
```suggestion
```

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+                          DEF LINK3_TRANSFORM HingeJoint {
+                            jointParameters HingeJointParameters {
+                              axis 0 0 1
+                              anchor 0 0 0
```
```suggestion
```

ad-daniel replied:
default value

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+                                  }
+                                ]
+                              }
+                              physics Physics {
```
The specs provide the densities? Not the masses?

ad-daniel commented on projects/robots/epson/ScaraT6/protos/ScaraEpsonT6.proto:
```diff
@@ -0,0 +1,626 @@
+  }
+  %< if (fields.staticBase.value) { >%
+  physics Physics {
+    density 1500
```
ditto

ad-daniel commented on docs/guide/scara_epson_t6.md:
```diff
@@ -0,0 +1,44 @@
+
+[SCARA Epson T6](https://www.epson.eu/products/robot/scara-t6-series) is a 4-axis robotic arm with three revolute and one prismatic DOF. The chosen model is the T6-602s. It can reach 200mm in vertical, 600mm in horizontal and can handle a payload of maximum 6kg. An end-effector can be choose on its shaft, as a gripper or a suction cup.
+
+The actuators available are three rotational motors (`base_arm_motor`, `arm_motor` and `shaft_rotation_motor`), a linear motor (`shaft_linear_motor`) and a [LEDs](../reference/led.md) (`epson_led`). The motors have four position sensors named `base_arm_position`, `arm_position` `shaft_rotation_position` and `shaft_linear_position`.
```
```suggestion
The actuators available are three rotational motors (`base_arm_motor`, `arm_motor` and `shaft_rotation_motor`), a linear motor (`shaft_linear_motor`) and a [LED](../reference/led.md) (`epson_led`).
The motors have four position sensors named `base_arm_position`, `arm_position` `shaft_rotation_position` and `shaft_linear_position`.
```

ad-daniel commented on docs/guide/scara_epson_t6.md:
```diff
@@ -0,0 +1,44 @@
+
+%robot SCARA Epson T6 images/robots/ScaraT6/model.png
+
+[SCARA Epson T6](https://www.epson.eu/products/robot/scara-t6-series) is a 4-axis robotic arm with three revolute and one prismatic DOF. The chosen model is the T6-602s. It can reach 200mm in vertical, 600mm in horizontal and can handle a payload of maximum 6kg. An end-effector can be choose on its shaft, as a gripper or a suction cup.
```
```suggestion
[SCARA Epson T6](https://www.epson.eu/products/robot/scara-t6-series) is a 4-axis robotic arm with three revolute and one prismatic DOF.
The chosen model is the T6-602s.
It has a 200mm vertical reach, a 600mm horizontal one and can handle a payload of maximum 6kg.
An end-effector can be choose on its shaft, as a gripper or a suction cup.
```

ad-daniel replied:
one sentence per line and suggested a change for the line about the reach

ad-daniel commented on docs/guide/scara_epson_t6.md:
```diff
@@ -0,0 +1,44 @@
+#### industrial\_example.wbt
+
+![industrial_example.wbt.png](images/robots/ScaraT6/industrial_example.wbt.png) In this example, you can see the SCARA robot moving fruits from a track to baskets using a suction cup.
+It is using the `scara_food_industry` controller.
```
```suggestion
```

ad-daniel replied:
Seems unnecessary to me

ad-daniel commented on projects/robots/epson/ScaraT6/worlds/.industrial_example.wbproj:
```diff
@@ -0,0 +1,11 @@
+renderingMode: PLAIN
+orthographicViewHeight: 1
+textFiles: 0 "controllers/scara_food_industry/scara_food_industry.py"
+globalOptionalRendering: JointAxes
```
Probably unnecessary to show the joint axis in the final demo

ad-daniel reviewed (changes requested):
Thank you, cool robot!
A couple of things:
- loading the world shows 2 warnings (unconnected field and oblong object)
- I can't seem to be able to open the doc for this robot (https://cyberbotics.com/doc/guide/scara_epson_t6?version=feature-proto-scara), does it work for you?
- The anchor point for a joint might be out of place
![industrial_example](https://user-images.githubusercontent.com/44834743/153360141-74f63fc6-e697-45e8-9503-ce9c45545173.png)

ad-daniel commented on docs/guide/robots.md:
```diff
@@ -54,6 +54,7 @@ Please [contact us](https://cyberbotics.com/#contact) if you would like to see y
 | ![icon.png](images/robots/robotis-op2/icon.png)      | [Robotis OP2](robotis-op2.md)             | Robotis              | *Small size humanoid*             |
 | ![icon.png](images/robots/robotis-op3/icon.png)      | [Robotis OP3](robotis-op3.md)             | Robotis              | *Small size humanoid*             |
 | ![icon.png](images/robots/salamander/icon.png)       | [Salamander](salamander.md)               | EPFL / BioRob        | *Bio-inspired robot*              |
+| ![icon.png](images/robots/scaraEpsonT6/icon.png)          | [ScaraT6](scara-epson-t6.md)              | Epson                | *SCARA robot*                     |
```
This line should be aligned as the rest: [raw text](https://raw.githubusercontent.com/cyberbotics/webots/53c51a5930d4572acd25cb39eb665e002ae5da30/docs/guide/robots.md)

 (might as well fix the one of the Ned robot as well)

ad-daniel commented on docs/guide/scara-epson-t6.md:
```diff
@@ -0,0 +1,50 @@
+[SCARA Epson T6](https://www.epson.eu/products/robot/scara-t6-series) is a 4-axis robotic arm with three revolute and one prismatic DOF.
+The chosen model is the T6-602s.
+It can reach 200mm in vertical with the shaft.
+Each arm has a radius of 300mm and can thus reach 600mm in horizontal.
```
```suggestion
Each arm has a radius of 300mm and can thus reach 600mm horizontally.
```

ad-daniel commented on docs/guide/scara-epson-t6.md:
```diff
@@ -0,0 +1,50 @@
+
+[SCARA Epson T6](https://www.epson.eu/products/robot/scara-t6-series) is a 4-axis robotic arm with three revolute and one prismatic DOF.
+The chosen model is the T6-602s.
+It can reach 200mm in vertical with the shaft.
```
```suggestion
It has a 200mm vertical range with the shaft.
```

ad-daniel replied:
Just a suggestion, the phrase sounds odd to my hear, dunno why. Up to you

ad-daniel commented on docs/guide/scara-epson-t6.md:
```diff
@@ -0,0 +1,50 @@
+It can reach 200mm in vertical with the shaft.
+Each arm has a radius of 300mm and can thus reach 600mm in horizontal.
+It can handle a payload of maximum 6kg.
+An end-effector can be choose on its shaft, as a gripper or a suction cup.
```
```suggestion
An end-effector can be chosen on its shaft, as a gripper or a suction cup.
```

ad-daniel commented on projects/robots/epson/scaraEpsonT6/protos/scaraEpsonT6.proto:
```diff
@@ -0,0 +1,632 @@
+# Model of the T6-602S Epson Scara robot.
+# template language: javascript
+
+PROTO scaraEpsonT6 [
```
```suggestion
PROTO ScaraEpsonT6 [
```

ad-daniel replied:
PROTO are always capitalized, also need to rename the file accordingly.

ad-daniel commented on projects/robots/epson/scaraEpsonT6/protos/scaraEpsonT6.proto:
```diff
@@ -0,0 +1,632 @@
+PROTO scaraEpsonT6 [
+  field SFVec3f           translation     0 0 0.01        # Is `Transform.translation`.
+  field SFRotation        rotation        0 0 1 0         # Is `Transform.rotation`.
+  field SFString          name            "scaraEpsonT6"  # Is `Solid.name`.
```
```suggestion
  field SFString          name            "ScaraEpsonT6"  # Is `Solid.name`.
```

ad-daniel commented on docs/guide/scara-epson-t6.md:
```diff
@@ -0,0 +1,50 @@
+scaraEpsonT6 [
+  SFVec3f           translation     0 0 0.01
+  SFRotation        rotation        0 0 1 0
+  SFString          name            "scaraEpsonT6"
```
```suggestion
  SFString          name            "ScaraEpsonT6"
```

ad-daniel commented on docs/guide/scara-epson-t6.md:
````diff
@@ -0,0 +1,50 @@
+Derived from [Robot](../reference/robot.md).
+
+```
+scaraEpsonT6 [
````
```suggestion
ScaraEpsonT6 [
```

ad-daniel replied:
Protos are always capitalized

ad-daniel commented on projects/robots/epson/scaraEpsonT6/protos/scaraEpsonT6.proto:
```diff
@@ -0,0 +1,632 @@
+  field SFString          window          ""              # Is `Robot.window`.
+  field SFBool            staticBase      FALSE           # Defines if the scara base should be pinned to the static environment.
+  field MFNode            handSlot        []              # Extends the shaft with new nodes in the hand slot.
+
```
```suggestion
```

ad-daniel commented on scripts/web_component_studio/components.json:
```diff
@@ -451,6 +451,16 @@
     },
+    {
+      "name": "scaraEpsonT6",
+      "proto": "scaraEpsonT6",
```
```suggestion
      "proto": "ScaraEpsonT6",
```

ad-daniel commented on projects/robots/epson/scaraEpsonT6/worlds/.industrial_example.wbproj:
```diff
@@ -0,0 +1,11 @@
+Webots Project File version R2022b
+perspectives: 000000ff00000000fd0000000200000001000002a400000285fc0200000001fb0000001400540065007800740045006400690074006f00720100000016000002850000008900ffffff000000030000073800000145fc0100000001fb0000001a0043006f006e0073006f006c00650041006c006c0041006c006c0100000000000007380000006900ffffff000004920000028500000001000000020000000100000008fc00000000
```
The default layout seems odd (too big node tree, too big console, too small 3d window)
![a](https://user-images.githubusercontent.com/44834743/154435853-a454f7ff-d044-4f70-9efb-6e1af5140315.png)

ad-daniel commented on projects/robots/epson/scaraEpsonT6/protos/scaraEpsonT6.proto:
```diff
@@ -0,0 +1,632 @@
+
+]
+{
+Robot {
```
is the anchor of the slider in the correct spot?
![industrial_example_1](https://user-images.githubusercontent.com/44834743/154436364-758d311d-3e02-4f61-a85e-25c7d37e86c6.png)

ad-daniel commented on projects/robots/epson/scaraEpsonT6/protos/scaraEpsonT6.proto:
```diff
@@ -0,0 +1,632 @@
+  translation IS translation
+  rotation IS rotation
+  synchronization IS synchronization
+  name IS name
```
The anchor of the first joint should be where the arm actually revolves instead of being on the ground

![industrial_example_2](https://user-images.githubusercontent.com/44834743/154436921-ad2162a3-a665-451a-91f7-ee8b5f57d730.png)

ad-daniel commented on projects/robots/epson/scaraEpsonT6/protos/scaraEpsonT6.proto:
```diff
@@ -0,0 +1,632 @@
+    density -1
+    mass 17.7 # estimated, calibrated on the total mass, 22kg
+    centerOfMass [
+      0 0 -0.01
```
Is it normal that the CoM is below ground?

ad-daniel reviewed (changes requested):
Thank you, there's still a few things that I think should be addressed

joachimhgg commented:
I re-work on the PROTO to improve the render of the web component and to change the position of the joints. I fixed the CoM as well. The axis of the joint were not corresponding to the axis of the meshes, hence it was needed to change the axis of the meshes. I fixed it for the arm because it was really not corresponding, but not for the base because the variation is very small and concern only the web component. Let me know if you think it could be better to do it as well.

ad-daniel reviewed (dismissed):
As discussed, the issue with the anchors appears to be on the web component side, hence this PR can be merged and an issue should be opened about it. 
![industrial_example_3](https://user-images.githubusercontent.com/44834743/154694974-059f379d-fa46-4ea0-bdbd-0cd50087671b.png)


Other than that, the rest looks fine to me.

omichel reviewed (changes requested):
It looks very nice!
A small but important detail: the paths in Webots are all lowercase, which means you should rename the `scaraT6` folders to `scara_t6` (two instances in the `docs` folder and one in the `robots` folder).

omichel commented on docs/guide/scara-epson-t6.md:
````diff
@@ -0,0 +1,50 @@
+
+```
+ScaraT6 [
+  SFVec3f           translation     0 0 0.01
````
```suggestion
  SFVec3f           translation     0 0 0
```

omichel reviewed (approved):
Thank you.
